### PR TITLE
Added parallel STFT implementation

### PR DESCRIPTION
**Pull request by JPery**
Hi!

As I comented in #98, I added a parallel STFT implementation based on the map_fn function following the indications of @zaccharieramzi [here](https://github.com/tensorflow/tensorflow/issues/6541#issuecomment-728947886).

I've added a `use_parallel_stft` param (disabled by default) that allows to use this function. I've put this param in as many functions as I can. I also added test cases for every function I can, including an specific test that checks that the result of the `tf.signal.stft` is equals to the result of the `parallel_stft` function.

Hope this could serve us well meanwhile tensorflow resolves its issues with the fft implementation.

### CODE REVIEW

File: kapre/backend.py [hunk 1]
```diff
@@ -10,6 +10,10 @@
 """
 from tensorflow.keras import backend as K
 import tensorflow as tf
+from tensorflow.python.ops.signal import shape_ops, fft_ops, window_ops, spectral_ops
+from tensorflow.python.framework import ops
+from functools import partial
+import multiprocessing
```
Comment (keunwoochoi) [inline kapre/backend.py:16]: I'd like importing `multiprocessing` before Tensorflow.

File: kapre/backend.py [hunk 2]
```diff
@@ -248,3 +252,35 @@ def mu_law_decoding(signal_mu, quantization_channels):
         tf.math.sign(signal) * (tf.math.exp(tf.math.abs(signal) * tf.math.log1p(mu)) - 1.0) / mu
     )
     return signal
+
+
+def parallel_stft(signals, frame_length, frame_step, fft_length=None, window_fn=window_ops.hann_window,
```
Comment (keunwoochoi) [inline kapre/backend.py:257]: `./script/apply-black.sh` would format things like this. Please run it once before pushing!

Comment (keunwoochoi) [reply kapre/backend.py:257]: Docstring please! Maybe we should clarify that it is an alternative funtion of `tf.signal.stft()` somewhere and that that's why we chose the names to be like these.

Comment (keunwoochoi) [reply kapre/backend.py:257]: How about adding `n_cpu=None` in the function signature and it defaults to be `multiprocessing.cpu_count()`?

File: kapre/composed.py [hunk 3]
```diff
@@ -44,6 +44,7 @@ def get_stft_magnitude_layer(
     input_data_format='default',
     output_data_format='default',
     name='stft_magnitude',
+    use_parallel_stft=False,
```
Comment (keunwoochoi) [inline kapre/composed.py:47]: We'd need this in the docstring, too, something like.. 
```
use_parallel_stft (bool): Whether to parallelize stft when running on CPU. If `True`, it uses `kapre.backend.parallel_stft()` which uses multi-processing. If `False`, it uses Tensorflow `tf.signal.stft`, which is significant slower than other STFT implementations such as librosa or scipy on CPU. It does not affect the behavior when running on GPUs. 
```

I think the variable name is good.

File: kapre/composed.py [hunk 4]
```diff
@@ -156,6 +158,7 @@ def get_melspectrogram_layer(
     input_data_format='default',
     output_data_format='default',
     name='melspectrogram',
+    use_parallel_stft=False
```
Comment (keunwoochoi) [inline kapre/composed.py:161]: same here (docstring)

File: kapre/composed.py [hunk 5]
```diff
@@ -234,6 +237,7 @@ def get_melspectrogram_layer(
         pad_end=pad_end,
         input_data_format=input_data_format,
         output_data_format=output_data_format,
+        use_parallel_stft=use_parallel_stft,
```
Comment (keunwoochoi) [inline kapre/composed.py:240]: same here (docstring)

File: kapre/composed.py [hunk 6]
```diff
@@ -281,6 +285,7 @@ def get_log_frequency_spectrogram_layer(
     input_data_format='default',
     output_data_format='default',
     name='log_frequency_spectrogram',
+    use_parallel_stft=False,
```
Comment (keunwoochoi) [inline kapre/composed.py:288]: same here (docstring)

File: kapre/composed.py [hunk 7]
```diff
@@ -349,6 +354,7 @@ def get_log_frequency_spectrogram_layer(
         pad_end=pad_end,
         input_data_format=input_data_format,
         output_data_format=output_data_format,
+        use_parallel_stft=use_parallel_stft,
```
Comment (keunwoochoi) [inline kapre/composed.py:357]: same here (docstring)

File: kapre/composed.py [hunk 8]
```diff
@@ -396,6 +402,7 @@ def get_perfectly_reconstructing_stft_istft(
     stft_data_format='default',
     stft_name='stft',
     istft_name='istft',
+    use_parallel_stft=False
```
Comment (keunwoochoi) [inline kapre/composed.py:405]: same here (docstring)

File: kapre/composed.py [hunk 9]
```diff
@@ -483,6 +490,7 @@ def get_perfectly_reconstructing_stft_istft(
         input_data_format=waveform_data_format,
         output_data_format=stft_data_format,
         name=stft_name,
+        use_parallel_stft=use_parallel_stft,
```
Comment (keunwoochoi) [inline kapre/composed.py:493]: same here (docstring)

File: kapre/composed.py [hunk 10]
```diff
@@ -514,6 +522,7 @@ def get_stft_mag_phase(
     input_data_format='default',
     output_data_format='default',
     name='stft_mag_phase',
+    use_parallel_stft=False
```
Comment (keunwoochoi) [inline kapre/composed.py:525]: same here (docstring)

File: kapre/composed.py [hunk 11]
```diff
@@ -566,6 +575,7 @@ def get_stft_mag_phase(
         pad_end=pad_end,
         input_data_format=input_data_format,
         output_data_format=output_data_format,
+        use_parallel_stft=use_parallel_stft,
```
Comment (keunwoochoi) [inline kapre/composed.py:578]: same here (docstring)

File: kapre/time_frequency.py [hunk 12]
```diff
@@ -105,6 +105,7 @@ def __init__(
         pad_end=False,
         input_data_format='default',
         output_data_format='default',
+        use_parallel_stft=False,
```
Comment (keunwoochoi) [inline kapre/time_frequency.py:108]: If we let users specify the number of processings, we'd need a variable for it here too. 
(This applies to all the places in `composed.py`, too)

File: kapre/time_frequency.py [hunk 13]
```diff
@@ -129,6 +130,8 @@ def __init__(
         self.output_data_format = K.image_data_format() if odt == _CH_DEFAULT_STR else odt
         self.input_data_format = K.image_data_format() if idt == _CH_DEFAULT_STR else idt
 
+        self.use_parallel_stft=use_parallel_stft
```
Comment (keunwoochoi) [inline kapre/time_frequency.py:133]: another formatting issue that `apply-black.sh` will fix.

File: kapre/time_frequency.py [hunk 14]
```diff
@@ -156,8 +159,12 @@ def call(self, x):
             waveforms = tf.pad(
                 waveforms, tf.constant([[0, 0], [0, 0], [int(self.n_fft - self.hop_length), 0]])
             )
+        
+        stft_function = tf.signal.stft
```
Comment (keunwoochoi) [inline kapre/time_frequency.py:163]: Can we make it even more straightforward like..
```
if self.use_parallel_stft:
    stft_function = parallel_stft
else:
    stft_function = tf.signal.stft
```
? 

Or an one-liner like `stft_function = parallel_stft if self.use_parallel_stft else tf.signal.stft`

File: tests/test_backend.py [hunk 15]
```diff
@@ -128,5 +128,21 @@ def test_validate_fail():
     _ = validate_data_format_str('weird_string')
 
 
+@pytest.mark.parametrize('frame_length', [1024, 2048])
+@pytest.mark.parametrize('frame_step', [256, 512])
+@pytest.mark.parametrize('image_size', [256, 512, 1024])
+def test_parallel_stft_should_be_equal_to_tf_stft_on_cpu(frame_length, frame_step, image_size):
```
Comment (keunwoochoi) [inline tests/test_backend.py:134]: I don't think it's a bad name, but for the consistency with other functions, I'd like `test_parallel_stft_correctness` slightly more.

**Review by keunwoochoi — CHANGES_REQUESTED** [review]
Thanks a lot for this work. This is a pretty strong PR and all the logics seem making sense to me. 
I left some comments but all are easy fixes.

File: kapre/backend.py (hunk 2 above)
Comment (JPery) [reply kapre/backend.py:257]: > How about adding `n_cpu=None` in the function signature and it defaults to be `multiprocessing.cpu_count()`?

The [tf.map_fn docs](https://www.tensorflow.org/api_docs/python/tf/map_fn) say that `the default value is 10`. I think it's a better approach to use the available cpu number instead.

**Comment by keunwoochoi** [comment]
Agreed that using cpu_count() is better than defaulting to 10.  What I meant was to let user control the number of process optionally.

File: kapre/backend.py (hunk 2 above)
Comment (keunwoochoi) [reply kapre/backend.py]: Sorry, forgot to comment. I think setting the default value to follow `cpu_count()` is a good decision. But what I asked was about letting user set the number of cpus to use manually.

But actually.. that would add another variable for every place. So.. please forget about it, we can add it anytime later.

**Review by keunwoochoi — APPROVED** [review]

**Comment by keunwoochoi** [comment]
@JPery Thanks a lot again! I'm merging it to a temporary branch and will ultimately merge it to the next release.

**Comment by oliskir** [comment]
Hi,

I have been testing Kapre's STFT layer against the spectrogram computation methods we have implemented in [Ketos](https://docs.meridian.cs.dal.ca/ketos/) which are based on numpy's rfft and tile functions. 

On CPU, I observe the same behaviour as noted previously by @JPery : Kapre is significantly slower than the numpy-based implementation, taking about 30 times longer to compute spectrograms. (For reference, I'm working on a Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz.)

I have tried switching to the parallel_stft branch and setting `use_parallel_stft=True` when I create the STFT layer, hoping that this would solve the issue. However, Kapre remains about a factor 30 slower. Is there something that I've misunderstood? I was under the impression that the fix implemented @JPery should lead to faster execution on CPU.

Note: When I run my script with `use_parallel_stft=True`, tensorflow prints the following warning/info message: 
```
WARNING:tensorflow:Setting parallel_iterations > 1 has no effect when executing eagerly. Consider calling map_fn with tf.function to execute fn in parallel.
```
Could this be the reason why I am not seeing any improvement?
I'm using tensorflow version 2.7.0.

Thanks,
Oliver

**Comment by JPery** [comment]
Hi, @oliskir,

Can you try to disable [TF eager execution](https://stackoverflow.com/questions/53429896/how-do-i-disable-tensorflows-eager-execution) to see if there is any difference?

**Comment by oliskir** [comment]
Thanks for the suggestion @JPery . Disabling eager execution did the trick:

audio duration: 3.0 s
sampling rate: 12000 Hz
window: 2400 samples,  step: 120 samples
spec shape:  (281, 1201)

## eager execution enabled:
WARNING:tensorflow:Setting parallel_iterations > 1 has no effect when executing eagerly. Consider calling map_fn with tf.function to execute fn in parallel.

| batch_size | ketos (s) | kapre (s) | ketos/kapre |
| ---------- | --------- | --------- | ----------- |
| 1 | 0.02 | 0.63 | 0.04 |
| 2 | 0.04 | 1.25 | 0.03 |
| 4 | 0.07 | 2.49 | 0.03 |
| 8 | 0.14 | 5.02 | 0.03 |
| 16 | 0.28 | 10.02 | 0.03 |

## eager execution disabled:

| batch_size | ketos (s) | kapre (s) | ketos/kapre |
| ---------- | --------- | --------- | ----------- |
| 1 | 0.02 | 0.11 | 0.21 |
| 2 | 0.04 | 0.11 | 0.37 |
| 4 | 0.07 | 0.11 | 0.61 |
| 8 | 0.15 | 0.12 | 1.29 |
| 16 | 0.28 | 0.12 | 2.34 |

**Pull request merged**